cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -1,55 +1,247 @@
+                        return Ok((transaction, receipt));
+                    }
+                }
+            }
+            GeneratorState::Complete(Err(e)) => return Err(e),
+            // By matching against the never type explicitly, we assert that the `Ok` value of the
+            // result is actually the never type and has not been changed since this
+            // line was written. The never type can never be constructed, so we cannot
+            // reach this line never anyway.
+            GeneratorState::Complete(Ok(never)) => match never {},
+        }
+    }
+}
+
+async fn matching_transaction_bloom_filter<C, F>(
+    connector: C,
+    start_of_swap: NaiveDateTime,
+    topics: Vec<Option<Topic>>,
+    action: &str,
+    matcher: F,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+    F: Fn(
+        Transaction,
+    ) -> Pin<
+        Box<dyn Future<Output = anyhow::Result<Option<(Transaction, TransactionReceipt)>>> + Send>,
+    >,
+{
+    let mut block_generator = Gen::new({
+        let connector = connector.clone();
+        |co| async move { find_relevant_blocks(connector, &co, start_of_swap).await }
+    });
+
+    'outer: loop {
+        match block_generator.async_resume().await {
+            GeneratorState::Yielded(block) => {
+                let block_hash = block
+                    .hash
+                    .ok_or_else(|| anyhow::anyhow!("block without hash"))?;
+
+                let maybe_contains_transaction = topics.iter().all(|topic| {
+                    topic.as_ref().map_or(true, |topic| {
+                        block
+                            .logs_bloom
+                            .contains_input(Input::Raw(topic.0.as_ref()))
+                    })
+                });
+                if !maybe_contains_transaction {
+                    tracing::trace!(
+                        "bloom filter indicates that block does not contain {} transaction:
+                {:x}",
+                        action,
+                        block_hash,
+                    );
                     continue;
                 }
+
+                tracing::trace!(
+                    "bloom filter indicates that we should check the block for {} transactions: {:x}",
+                    action,
+                    block_hash,
+                );
+                for transaction in block.transactions.into_iter() {
+                    if let Some((transaction, receipt)) = matcher(transaction.clone()).await? {
+                        // We can get a transaction that matches even for failed attempts
+                        // at contract creation e.g., low gas.
+                        if receipt.status != Some(TRANSACTION_STATUS_OK.into()) {
```

Comment:
We are doing this check in two different places. Might be worth defining a trait for it an implementing it on `TransactionReceipt`.